modules/tinymce/src/core/main/ts/fmt/FormatChanged.ts

```diff
@@ -5,101 +5,167 @@
-  const currentFormats = Cell<Record<string, FormatChangeCallback[]>>({ });
-
   registeredFormatListeners.set({});
 
   editor.on('NodeChange', (e) => {
-    updateAndFireChangeCallbacks(editor, e.element, currentFormats, registeredFormatListeners.get());
+    updateAndFireChangeCallbacks(editor, e.element, registeredFormatListeners.get());
   });
 };
 
-const updateAndFireChangeCallbacks = (editor: Editor, elm: Element, currentFormats: Cell<FormatCallbacks>, formatChangeData: RegisteredFormats) => {
-  const formatsList = Obj.keys(currentFormats.get());
-  const newFormats: FormatCallbacks = { };
-  const matchedFormats: FormatCallbacks = { };
+const matchingNode = (editor: Editor, parents: Element[], format: string, similar: boolean, vars?: FormatVars): Optional<Element> =>
+  Arr.findUntil(
+    parents,
+    (node) => { // is this the node we want?
+      const matchingFormat = editor.formatter.matchNode(node, format, vars ?? {}, similar);
+      return !Type.isUndefined(matchingFormat);
+    },
+    (node) => { // Can we guarantee that no parent of this will be the node we want?
+      if (MatchFormat.matchesUnInheritedFormatSelector(editor, node, format)) {
+        return true;
+      } else {
+        if (!similar) {
+          // If we want to find an exact match, then finding a similar match halfway up the parents tree is bad
+          return Type.isNonNullable(editor.formatter.matchNode(node, format, vars, true));
+        } else {
+          return false;
+        }
+      }
+    }
+  );
+
+const getParents = (editor: Editor, elm?: Element): Element[] => {
+  const element = elm ?? editor.selection.getNode();
+  return Arr.filter(FormatUtils.getParents(editor.dom, element), (node) =>
+    node.nodeType === 1 && !node.getAttribute('data-mce-bogus')
+  );
+};
+
+const hasChanged = (state: Cell<boolean>, newState: boolean): boolean => {
+  if (state.get() !== newState) {
+    state.set(newState);
+    return true;
+  } else {
+    return false;
+  }
+};
 
+const updateAndFireChangeCallbacks = (
+  editor: Editor,
+  elm: Element,
+  registeredCallbacks: RegisteredFormats
+) => {
   // Ignore bogus nodes like the <a> tag created by moveStart()
-  const parents = Arr.filter(FormatUtils.getParents(editor.dom, elm), (node) => node.nodeType === 1 && !node.getAttribute('data-mce-bogus'));
-
-  // Check for new formats
-  Obj.each(formatChangeData, (data: FormatData, format: string) => {
-    Tools.each(parents, (node: Node) => {
-      if (editor.formatter.matchNode(node, format, {}, data.similar)) {
-        if (formatsList.indexOf(format) === -1) {
-          // Execute callbacks
-          Arr.each(data.callbacks, (callback: FormatChangeCallback) => {
-            callback(true, { node, format, parents });
-          });
-
-          newFormats[format] = data.callbacks;
-        }
+  const parents = getParents(editor, elm);
+
+  Obj.each(registeredCallbacks, (data: CallbackGroup, format: string) => {
```

Comment:
```suggestion
  Obj.each(registeredCallbacks, (data, format) => {
```
The types can be inferred here so we shouldn't be including them.